File: modules/spike-metrics/ios/SpikeMetricsModule.swift
import ExpoModulesCore
import Foundation
import UIKit

internal struct DeviceProfile: Record {
  @Field var label: String = ""
  @Field var model: String = ""
  @Field var soc: String = ""
  @Field var osVersion: String = ""
  @Field var abi: String = ""
  @Field var cpuCores: Int = 0
  @Field var totalRamMb: Int = 0
}

internal struct SourceInfo: Record {
  @Field var name: String = ""
  @Field var sizeBytes: Double = 0
  @Field var readable: Bool = false
}

/// Throwaway Phase 0 instrumentation. Delete this module together with `spike/`
/// once the Stage 0 accuracy question is answered.
public class SpikeMetricsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("SpikeMetrics")

    Function("getDeviceProfile") { () -> DeviceProfile in
      let identifier = machineIdentifier()
      var profile = DeviceProfile()
      profile.model = identifier
      profile.soc = identifier
      profile.osVersion = "iOS \(UIDevice.current.systemVersion)"
      profile.abi = "arm64"
      profile.cpuCores = ProcessInfo.processInfo.processorCount
      profile.totalRamMb = Int(ProcessInfo.processInfo.physicalMemory / (1024 * 1024))
      profile.label = "\(identifier) / arm64"
      return profile
    }

    /// iOS has no VmHWM equivalent, so this reports the current phys_footprint and
    /// the caller keeps its own running maximum.
    Function("getPeakRssBytes") { () -> Double in
      Double(physFootprintBytes())
    }

    Function("getCurrentRssBytes") { () -> Double in
      Double(physFootprintBytes())
    }

    /// Nothing to reset on iOS. Returning false tells the rig to treat the numbers
    /// as sampled rather than as a kernel-maintained watermark.
    Function("resetPeakRss") { () -> Bool in
      false
    }

    /// Name and size behind a picked URI, plus whether it can still be opened.
    ///
    /// The picker hands back a copy in a temporary directory whose last path
    /// component is already the real filename, so unlike Android there is no
    /// provider to ask. `readable` matters all the same: a remembered clip in a
    /// temporary directory may have been swept away between runs.
    AsyncFunction("describeSource") { (uri: String) -> SourceInfo in
      var info = SourceInfo()
      guard let url = URL(string: uri) else { return info }

      info.name = url.lastPathComponent.removingPercentEncoding ?? url.lastPathComponent
      guard url.isFileURL else { return info }

      let attributes = try? FileManager.default.attributesOfItem(atPath: url.path)
      info.sizeBytes = Double((attributes?[.size] as? NSNumber)?.int64Value ?? 0)
      info.readable = FileManager.default.isReadableFile(atPath: url.path)
      return info
    }

    /// Mirrors the Android `Caption` logcat tag so a release build can be read from
    /// the device console, where JS `console.log` is not dependable.
    Function("log") { (message: String) in
      NSLog("Caption: %@", message)
    }
  }
}

private func machineIdentifier() -> String {
  var systemInfo = utsname()
  uname(&systemInfo)
  let mirror = Mirror(reflecting: systemInfo.machine)
  return mirror.children.reduce(into: "") { identifier, element in
    guard let value = element.value as? Int8, value != 0 else { return }
    identifier += String(UnicodeScalar(UInt8(value)))
  }
}

private func physFootprintBytes() -> UInt64 {
  var info = task_vm_info_data_t()
  var count = mach_msg_type_number_t(MemoryLayout<task_vm_info_data_t>.size / MemoryLayout<natural_t>.size)
  let result = withUnsafeMutablePointer(to: &info) {
    $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
      task_info(mach_task_self_, task_flavor_t(TASK_VM_INFO), $0, &count)
    }
  }
  return result == KERN_SUCCESS ? info.phys_footprint : 0
}

File: modules/spike-metrics/android/src/main/java/expo/modules/spikemetrics/SpikeMetricsModule.kt
package expo.modules.spikemetrics

import android.app.ActivityManager
import android.content.Context
import android.net.Uri
import android.os.Build
import android.provider.OpenableColumns
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.io.File

class DeviceProfile : Record {
  @Field var label: String = ""
  @Field var model: String = ""
  @Field var soc: String = ""
  @Field var osVersion: String = ""
  @Field var abi: String = ""
  @Field var cpuCores: Int = 0
  @Field var totalRamMb: Int = 0
}

class SourceInfo : Record {
  @Field var name: String = ""
  @Field var sizeBytes: Double = 0.0
  @Field var readable: Boolean = false
}

/**
 * Throwaway Phase 0 instrumentation. Delete this module together with `spike/`
 * once the Stage 0 accuracy question is answered; nothing in the shipped app
 * should depend on it.
 */
class SpikeMetricsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("SpikeMetrics")

    Function("getDeviceProfile") {
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      val memoryInfo = ActivityManager.MemoryInfo()
      (context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager)
        .getMemoryInfo(memoryInfo)

      val soc = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        "${Build.SOC_MANUFACTURER} ${Build.SOC_MODEL}".trim()
      } else {
        Build.HARDWARE
      }

      DeviceProfile().apply {
        model = "${Build.MANUFACTURER} ${Build.MODEL}".trim()
        this.soc = soc
        osVersion = "Android ${Build.VERSION.RELEASE} (API ${Build.VERSION.SDK_INT})"
        abi = Build.SUPPORTED_ABIS.firstOrNull() ?: "unknown"
        cpuCores = Runtime.getRuntime().availableProcessors()
        totalRamMb = (memoryInfo.totalMem / (1024 * 1024)).toInt()
        label = "$model / $soc / $abi"
      }
    }

    /**
     * Peak resident set size since process start, or since the last successful
     * [resetPeakRss], read from VmHWM in /proc/self/status. Returns 0 when the
     * kernel does not expose it.
     */
    Function("getPeakRssBytes") {
      readProcStatusKb("VmHWM:")?.let { it * 1024L } ?: 0L
    }

    Function("getCurrentRssBytes") {
      readProcStatusKb("VmRSS:")?.let { it * 1024L } ?: 0L
    }

    /**
     * Asks the kernel to reset the VmHWM watermark so each model run reports its
     * own peak rather than the high water mark of every run before it. Returns
     * false when the kernel refuses, in which case peaks are cumulative and the
     * caller should say so.
     */
    Function("resetPeakRss") {
      try {
        File("/proc/self/clear_refs").writeText("5")
        true
      } catch (_: Throwable) {
        false
      }
    }

    /**
     * Real display name and size behind a picked `content://` URI, plus whether it
     * can still be opened.
     *
     * A document URI carries a provider id, not a filename, so the picked clip
     * would otherwise land in the CSV as something like `msf:1000000045`. Only the
     * provider knows the name, and it only answers over the resolver. `readable`
     * reports whether a remembered URI still works: SAF grants survive a restart,
     * but not the file being deleted or the grant being revoked.
     */
    AsyncFunction("describeSource") { uri: String ->
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      val parsed = Uri.parse(uri)
      val info = SourceInfo()

      if (parsed.scheme == "content") {
        try {
          context.contentResolver
            .query(parsed, arrayOf(OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE), null, null, null)
            ?.use { cursor ->
              if (cursor.moveToFirst()) {
                cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                  .takeIf { it >= 0 && !cursor.isNull(it) }
                  ?.let { info.name = cursor.getString(it) }
                cursor.getColumnIndex(OpenableColumns.SIZE)
                  .takeIf { it >= 0 && !cursor.isNull(it) }
                  ?.let { info.sizeBytes = cursor.getLong(it).toDouble() }
              }
            }
          // A query can be answered out of a cache after the grant is gone, so
          // readability is decided by actually opening the thing.
          context.contentResolver.openInputStream(parsed)?.use { info.readable = true }
        } catch (_: Throwable) {
          info.readable = false
        }
        return@AsyncFunction info
      }

      val file = File(parsed.path ?: uri)
      info.name = file.name
      info.sizeBytes = file.length().toDouble()
      info.readable = file.canRead()
      info
    }

    /**
     * Writes to the `Caption` logcat tag so a release build can be read with
     * `adb logcat -s RNWhisper:* Caption:*`. JS `console.log` is not dependable
     * once the bundle is minified, and the Stage 0 numbers only count in release.
     */
    Function("log") { message: String ->
      // logcat drops anything past roughly 4 kB in a single entry.
      message.chunked(3000).forEach { chunk -> android.util.Log.i("Caption", chunk) }
    }
  }

  private fun readProcStatusKb(key: String): Long? = try {
    File("/proc/self/status").useLines { lines ->
      lines.firstOrNull { it.startsWith(key) }
        ?.substringAfter(key)
        ?.trim()
        ?.removeSuffix(" kB")
        ?.trim()
        ?.toLongOrNull()
    }
  } catch (_: Throwable) {
    null
  }
}
